torch/csrc/autograd/custom_function.cpp

```diff
@@ -78,7 +85,42 @@ std::vector<Variable> _wrap_outputs(const std::unordered_set<at::TensorImpl*> &i
   return outputs;
 }
 
+template<class T>
+void CFunction<T>::apply(variable_list&& input_vars) {
+  bool is_executable =  GradMode::is_enabled() && any_variable_requires_grad(input_vars);
+  auto next_edges = collect_next_edges(input_vars);
+
+  std::shared_ptr<CustomFunc<T>> grad_fn = std::make_shared<CustomFunc<T>>();
+  grad_fn->ctx.set_next_edges(std::move(next_edges));
+  grad_fn->ctx.clear_input_metadata();
+
+  variable_list outputs;
+  {
+    AutoGradMode grad_mode(false);
+    outputs = T::forward(&grad_fn->ctx, input_vars);
+  }
+
+  _wrap_outputs(input_vars, grad_fn.ctx->non_differentiable, grad_fn.ctx->dirty_inputs, outputs, is_executable ? grad_fn : nullptr);
+}
+
+template<class T>
+variable_list CustomFunc<T>::apply(variable_list&& inputs) {
+  auto outputs = T::backward(&ctx, inputs);
 
+  if (outputs.size() != inputs.size()) {
```

Comment:
This check here differs from the python function implementation (which accepts a longer outputs list as long as the later outputs are undefined.) But I don't know why we did that in the first place, so the stricter check here seems fine. (@apaszke, maybe you know?)

Reply:
I suspect it's because we want to support the following use case:
https://github.com/pytorch/pytorch/blob/425d28c30abcac22d9bdd2d01cc83e3e0470eb11/test/test_autograd.py#L2131-L2138
Here there are some inputs to the forward function that don't need to participate in the autograd graph, but we still want the number of outputs from the custom backward function to match the number of inputs into the custom forward function.

Reply:
Also it would be great to minimize the behavioral difference between Python and C++ API, so that it's easier for people to convert their code to C++ :D